### DISCUSSION TITLE
Reference update times of photon from Nominatim

### DISCUSSION THREAD

**Discussion by hannesaddec** [komoot/photon#534] [Q&A]
We observe larger update times for 2 languages only as we try to process all EU langauges than reported in the documentation: 

Setup: Nominatim version,3.5.1,  machine size:  4 cores, 28 Gb RAM), SSD DISK, global extent
Setup: Photon version, 0.3.4,  , machine size: 4 cores 14 Gb. RAM). SSD DISK
Command: java -jar /opt/data/photon/photon-0.3.4.jar -nominatim-import -host 10.0.1.17 -port 5432 -database nominatim -user nominatim -password XXXXXX -languages **en,es**
However this lead to only 25% CPU load with 200 documents/sec processed over 3 days and producing a 26GB index. JAVA/Nominatim/Photon uses only on thread each. We never reached/peaked at RAM/CPU.

Any further advice besides to upgrade Nominatim to 3.6 and photon 0.3.4 ?

**Comment by lonvia** [comment]
Make sure you have tuned Postgresql as per https://nominatim.org/release-docs/latest/admin/Installation/#tuning-the-postgresql-database. How is the load on the Nominatim machine?

**Reply by hannesaddec** [reply]
load on nominatim is at 25% as well.. checking the tuning parameters..

**Reply by hannesaddec** [reply]
Configuration is as in the documentation except that Synchronous commit is ON. no transactions. Synchronous commit is about transactions. Reading database -> no transactions

**Reply by lonvia** [reply]
Reading databases is transactional as well. You want to get consistent data after all. However, it sounds like your problem is with the connection between the two machines. I don't have any experience with running this setup on different machines, so I can't be of much help there I'm afraid. Maybe somebody else here can.

**Reply by hannesaddec** [reply]
Thx for the advice- will test it out and will provide feedback.

**Reply by alfmarcua** [reply]
We have tried in a single machine with 4 cores, 48GB RAM, SDD DISK, global extent and results were almost the same, 90 hours which is more than 3 days. Load on nominatim database was also low. Java process took 1 core and it was not using all the RAM available by far. Java process take more than 40% of the total time which is about 36 hours. Hence we are not sure how could we speed up the process in order to take less than those 36 hours. Could have we miss some Java additional configuration?

**Reply by lonvia** [reply]
What does the following SQL EXPLAIN say in your Nominatim database:
```
EXPLAIN SELECT * FROM placex p, place_addressline pa
WHERE p.place_id = pa.address_place_id and pa.place_id = 12345678
and pa.cached_rank_address > 4 and pa.address_place_id != 12345678 and pa.isaddress
ORDER BY rank_address desc, fromarea desc, distance asc, rank_search desc;
```

**Reply by alfmarcua** [reply]
Those queries, although tiny (1 ms on average), take the most of the time (about 38% of the total time).

```
Gather Merge  (cost=12552.31..12622.92 rows=614 width=549)
  Workers Planned: 1
  ->  Sort  (cost=11552.30..11553.83 rows=614 width=549)
        Sort Key: p.rank_address DESC, pa.fromarea DESC, pa.distance, p.rank_search DESC
        ->  Nested Loop  (cost=45.16..11523.86 rows=614 width=549)
              ->  Parallel Bitmap Heap Scan on place_addressline pa  (cost=44.59..6251.14 rows=614 width=28)
                    Recheck Cond: (place_id = 12345678)
                    Filter: (isaddress AND (cached_rank_address > 4) AND (address_place_id <> 12345678))
                    ->  Bitmap Index Scan on idx_place_addressline_place_id  (cost=0.00..44.33 rows=1567 width=0)
                          Index Cond: (place_id = 12345678)
              ->  Index Scan using idx_place_id on placex p  (cost=0.57..8.59 rows=1 width=521)
                    Index Cond: (place_id = pa.address_place_id)

**Reply by alfmarcua** [reply]
<body>


<div class="container-fluid main-container">




<div id="header">



<h1 class="title toc-ignore">Query analysis</h1>
Here it is a query analysis of an interval of 93.5 minutes during an import process.


</div>


<div id="data-logging" class="section level2">
<h2>Data logging</h2>
<p>Time to start recording <strong>all</strong> queries: 2021-03-06 21:37:25</p>
<p>Time to finish recording <strong>all</strong> queries: 2021-03-06 23:10:55</p>
<p>Time elapsed: 93.5 minutes</p>
<p>During that time 12214704 queries have been recorded</p>
</div>
<div id="data-summary" class="section level2">
<h2>Data summary</h2>
<pre><code>## `summarise()` has grouped output by &#39;operation&#39;. You can override using the `.groups` argument.</code></pre>
<table>
<thead>
<tr class="header">
<th align="left">operation</th>
<th align="left">sqlKeyword</th>
<th align="right">avg</th>
<th align="right">totalSecs</th>
<th align="right">totalMins</th>
<th align="right">totalHours</th>
<th align="right">totalPct</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">execute fetch</td>
<td align="left">SELECT</td>
<td align="right">1552.2419231</td>
<td align="right">20.179145</td>
<td align="right">0.3363191</td>
<td align="right">0.0056053</td>
<td align="right">0.36</td>
</tr>
<tr class="even">
<td align="left">execute</td>
<td align="left">SELECT</td>
<td align="right">1.0192581</td>
<td align="right">1778.564639</td>
<td align="right">29.6427440</td>
<td align="right">0.4940457</td>
<td align="right">31.70</td>
</tr>
<tr class="odd">
<td align="left">execute</td>
<td align="left">ROLLBACK</td>
<td align="right">0.0175172</td>
<td align="right">30.566667</td>
<td align="right">0.5094444</td>
<td align="right">0.0084907</td>
<td align="right">0.54</td>
</tr>
<tr class="even">
<td align="left">execute</td>
<td align="left">BEGIN</td>
<td align="right">0.0032470</td>
<td align="right">5.665857</td>
<td align="right">0.0944309</td>
<td align="right">0.0015738</td>
<td align="right">0.10</td>
</tr>
<tr class="odd">
<td align="left">bind</td>
<td align="left">SELECT</td>
<td align="right">0.2099698</td>
<td align="right">366.387825</td>
<td align="right">6.1064638</td>
<td align="right">0.1017744</td>
<td align="right">6.53</td>
</tr>
<tr class="even">
<td align="left">bind</td>
<td align="left">ROLLBACK</td>
<td align="right">0.0046953</td>
<td align="right">8.193050</td>
<td align="right">0.1365508</td>
<td align="right">0.0022758</td>
<td align="right">0.15</td>
</tr>
<tr class="odd">
<td align="left">bind</td>
<td align="left">BEGIN</td>
<td align="right">0.0065285</td>
<td align="right">11.391967</td>
<td align="right">0.1898661</td>
<td align="right">0.0031644</td>
<td align="right">0.20</td>
</tr>
<tr class="even">
<td align="left">bind</td>
<td align="left">NA</td>
<td align="right">0.0067176</td>
<td align="right">11.721876</td>
<td align="right">0.1953646</td>
<td align="right">0.0032561</td>
<td align="right">0.21</td>
</tr>
</tbody>
</table>
</div>
<div id="time-by-sql-operation" class="section level2">
<h2>Time by SQL operation:</h2>
<table>
<thead>
<tr class="header">
<th align="left">operation</th>
<th align="right">avg</th>
<th align="right">total</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">execute fetch</td>
<td align="right">1552.2419231</td>
<td align="right">0.3363191</td>
</tr>
<tr class="even">
<td align="left">execute</td>
<td align="right">0.3466747</td>
<td align="right">30.2466194</td>
</tr>
<tr class="odd">
<td align="left">bind</td>
<td align="right">0.0569778</td>
<td align="right">6.6282453</td>
</tr>
</tbody>
</table>
</div>
<div id="time-by-query" class="section level2">
<h2>Time by QUERY</h2>
<table>
<colgroup>
<col width="91%" />
<col width="2%" />
<col width="3%" />
<col width="2%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">sqlClause</th>
<th align="right">nOcurr</th>
<th align="right">totalSecs</th>
<th align="right">totalPct</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">BEGIN</td>
<td align="right">1744955</td>
<td align="right">17.05782</td>
<td align="right">0.30</td>
</tr>
<tr class="even">
<td align="left">ROLLBACK</td>
<td align="right">1744955</td>
<td align="right">38.75972</td>
<td align="right">0.69</td>
</tr>
<tr class="odd">
<td align="left">SELECT p.place_id, p.name, p.class, p.type, p.rank_address 

FROM placex p WHERE p.place_id = $1</td>
<td align="right">606172</td>
<td align="right">407.82855</td>
<td align="right">7.27</td>
</tr>
<tr class="even">
<td align="left">SELECT p.place_id, p.name, p.class, p.type, p.rank_address

 FROM placex p, place_addressline pa WHERE p.place_id = pa.address_place_id 

and pa.place_id = $1 and pa.cached_rank_address &gt; 4 and 

pa.address_place_id != $2 and pa.isaddress order by rank_address desc,fromarea desc,

distance asc,rank_search desc</td>
<td align="right">1138786</td>
<td align="right">1737.12391</td>
<td align="right">30.96</td>
</tr>
<tr class="odd">
<td align="left">SELECT place_id, osm_type, osm_id, class, type, name, 

housenumber, postcode, address, extratags, ST_Envelope(geometry) AS bbox, 

parent_place_id, linked_place_id, rank_address, rank_search, importance, country_code, 

centroid FROM placex WHERE linked_place_id IS NULL AND centroid IS NOT NULL 
ORDER BY geometry_sector</td>
<td align="right">6</td>
<td align="right">20.17914</td>
<td align="right">0.36</td>
</tr>
<tr class="even">
<td align="left">NA</td>
<td align="right">872478</td>
<td align="right">11.72188</td>
<td align="right">0.21</td>
</tr>
</tbody>
</table>
</div>




</div>

**Reply by lonvia** [reply]
Looks like the statistics are off. It shouldn't be doing a heap scan on place_addressline. This is what the explain looks like for me:
```
 Sort  (cost=6452.10..6453.33 rows=494 width=567)
   Sort Key: p.rank_address DESC, pa.fromarea DESC, pa.distance, p.rank_search DESC
   ->  Nested Loop  (cost=1.14..6430.00 rows=494 width=567)
         ->  Index Scan using idx_place_addressline_place_id on place_addressline pa  (cost=0.57..2187.77 rows=494 width=28)
               Index Cond: (place_id = 12345678)
               Filter: (isaddress AND (cached_rank_address > 4) AND (address_place_id <> 12345678))
         ->  Index Scan using idx_place_id on placex p  (cost=0.57..8.59 rows=1 width=539)
               Index Cond: (place_id = pa.address_place_id)
```

Try running a VACUUM ANALYSE on place_addressline. If that doesn't help set `random_page_cost = 1.0` in the postgresql configuration.

**Comment by enriquezubiria** [comment]
@lonvia , please keep in mind that the numbers about average and total query time that my colleague Alfonso has sent to you show only a limited window of time while we have had enabled full logging on our database.

**Comment by enriquezubiria** [comment]
Sarah, thanks for your help.

That is exactly what we were doing. In fact, Alfonso has run the EXPLAIN
while a VACUUM ANALYZE was being run.

I'll rerun the EXPLAIN when the vacuum command finishes and I'll send you
the results.

Thanks again,

Enrique Z.

**Reply by enriquezubiria** [reply]
@lonvia , here are the results of running the explain command you asked for after having run the vacuum analyze:
```
nominatim=# EXPLAIN SELECT * FROM placex p, place_addressline pa
nominatim-# WHERE p.place_id = pa.address_place_id and pa.place_id = 12345678
nominatim-# and pa.cached_rank_address > 4 and pa.address_place_id != 12345678 and pa.isaddress
nominatim-# ORDER BY rank_address desc, fromarea desc, distance asc, rank_search desc;
QUERY PLAN
-------------------------------------------------------------------------------------------------------------------
Gather Merge (cost=12552.31..12622.92 rows=614 width=640)
Workers Planned: 1
-> Sort (cost=11552.30..11553.83 rows=614 width=640)
Sort Key: p.rank_address DESC, pa.fromarea DESC, pa.distance, p.rank_search DESC
-> Nested Loop (cost=45.16..11523.86 rows=614 width=640)
-> Parallel Bitmap Heap Scan on place_addressline pa (cost=44.59..6251.14 rows=614 width=28)
Recheck Cond: (place_id = 12345678)
Filter: (isaddress AND (cached_rank_address > 4) AND (address_place_id <> 12345678))
-> Bitmap Index Scan on idx_place_addressline_place_id (cost=0.00..44.33 rows=1567 width=0)
Index Cond: (place_id = 12345678)
-> Index Scan using idx_place_id on placex p (cost=0.57..8.59 rows=1 width=612)
Index Cond: (place_id = pa.address_place_id)
(12 rows)
```
Thanks

**Reply by lonvia** [reply]
Still the heap scan. Have you tried with the random_page_cost option? Change the postgresql config, reload postgresql, rerun EXPLAIN.

**Reply by enriquezubiria** [reply]
After changing `random_page_cost` the result seems better:
```
 Sort  (cost=4271.04..4273.65 rows=1043 width=640)
   Sort Key: p.rank_address DESC, pa.fromarea DESC, pa.distance, p.rank_search DESC
   ->  Nested Loop  (cost=1.15..4218.75 rows=1043 width=640)
         ->  Index Scan using idx_place_addressline_place_id on place_addressline pa  (cost=0.57..1519.99 rows=1043 width=28)
               Index Cond: (place_id = 12345678)
               Filter: (isaddress AND (cached_rank_address > 4) AND (address_place_id <> 12345678))
         ->  Index Scan using idx_place_id on placex p  (cost=0.57..2.59 rows=1 width=612)
               Index Cond: (place_id = pa.address_place_id)
```

**Reply by lonvia** [reply]
Indeed. Lets see if import is going faster now.

**Reply by bt90** [reply]
Could you also post an `EXPLAIN (ANALYZE)` ? Would be interesting to see where the planer has bad estimates.

**Reply by alfmarcua** [reply]
@bt90 Here you have the results of the `EXPLAIN (ANALYZE)`:
```
Gather Merge  (cost=4454.78..4530.56 rows=659 width=640) (actual time=5.123..7.498 rows=0 loops=1)
  Workers Planned: 1
  Workers Launched: 1
  ->  Sort  (cost=3454.77..3456.41 rows=659 width=640) (actual time=0.273..0.274 rows=0 loops=2)
        Sort Key: p.rank_address DESC, pa.fromarea DESC, pa.distance, p.rank_search DESC
        Sort Method: quicksort  Memory: 25kB
        ->  Nested Loop  (cost=22.03..3423.91 rows=659 width=640) (actual time=0.237..0.238 rows=0 loops=2)
              ->  Parallel Bitmap Heap Scan on place_addressline pa  (cost=21.46..1718.75 rows=659 width=28) (actual time=0.237..0.237 rows=0 loops=2)
                    Recheck Cond: (place_id = 12345678)
                    Filter: (isaddress AND (cached_rank_address > 4) AND (address_place_id <> 12345678))
                    ->  Bitmap Index Scan on idx_place_addressline_place_id  (cost=0.00..21.18 rows=1680 width=0) (actual time=0.015..0.015 rows=0 loops=1)
                          Index Cond: (place_id = 12345678)
              ->  Index Scan using idx_place_id on placex p  (cost=0.57..2.59 rows=1 width=612) (never executed)
                    Index Cond: (place_id = pa.address_place_id)
Planning time: 0.431 ms
Execution time: 7.562 ms
```

**Reply by alfmarcua** [reply]
@lonvia Unfortunately, the import is not faster.
Could the PostgreSQL version of the nominating database has some impact on it? We are currently using PostgreSQL 10.15 with PostGIS 2.4.

**Reply by lonvia** [reply]
I'm out of ideas. I'd need to see the setup for myself.

**Reply by otbutz** [reply]
```
Bitmap Index Scan on idx_place_addressline_place_id  (cost=0.00..21.18 rows=1680 width=0) (actual time=0.015..0.015 rows=0 loops=1)
```

That's a pretty bad estimation. @lonvia are `place_addressline.isaddress` and `place_addressline.cached_rank_address` correlated?

**Reply by lonvia** [reply]
No, they are not correlated. What is likely correlated is the distribution of the number of rows per place_id and is_address. And what is definitely correlated is the number of rows per address_place_id and cached_rank_address.

**Reply by otbutz** [reply]
We could check if a planer hint would improve the estimation: https://www.postgresql.org/docs/10/sql-createstatistics.html (available since PostgreSQL 10)

**Reply by lonvia** [reply]
We'd have to confirm first that the query planner and that particular query is indeed the root cause of the slow export.

**Reply by alfmarcua** [reply]
According to the information we were handling with @hannesaddec, the estimated time for uploading nominatim to photon was about 19 hours for a planet export with a 6-core machine with 64GB RAM and SSD, while you have obtained about 7-8 hours in the last migration, see #547.

Is there any misunderstanding from our side in this information?

**Reply by alfmarcua** [reply]
> I'm out of ideas. I'd need to see the setup for myself.

Thanks a lot for your kind support, Sarah.

We have deployed a specific environment which is prepared to check the setup in case it is possible for you, as we are still getting long times for the export (even we are using the latest version which also includes a remarkable performance improvement https://github.com/komoot/photon/pull/547).

If you see this appropriate, we can provide you with more details to access this environment.
